src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs

```diff
@@ -1109,21 +1108,46 @@ internal void PrepareContainerForIndex(DependencyObject container, int index)
 
 		/// <summary>
 		/// Create a new container for item at <paramref name="index"/> and returns it, unless the item is its own container, in which case
-		/// returns the item itself.
+		/// returns the item itself, or the template root is a suitable container, in which case the template root is returned.
 		/// </summary>
+		/// <remarks>For virtualized controls, this method is only called if it's not possible to recycle an existing container.</remarks>
 		internal DependencyObject GetContainerForIndex(int index)
 		{
 			var item = ItemFromIndex(index);
 
-			var container = IsItemItsOwnContainerOverride(item)
-				? item as DependencyObject
-				: GetContainerForItemOverride();
+			if (IsItemItsOwnContainerOverride(item))
+			{
+				var container = item as DependencyObject;
+				container.SetValue(ItemsControlForItemContainerProperty, new WeakReference<ItemsControl>(this));
```

Comment:
You should be able to use `IWeakReferenceProvider.WeakReference` here, but maybe that's for another PR :)